### DISCUSSION TITLE
Action works in development but not in production

### DISCUSSION THREAD

**Discussion by thanosoncode** [vercel/next.js#50371] [App Router]
Any ideas why this deletePost action works in development but not in production?
In production I'm getting a `ReferenceError: id is not defined`. How the DeleteForm does not know about id?

```
const Home = async () => {
  const posts = (await prisma.post.findMany()) as Post[];

  return (
    <div>
      <h1 className="text-2xl mb-8">All posts</h1>
      <section className="flex flex-col gap-3">
        {posts.map((post) => (
          <div key={post.id} className="flex gap-4 justify-between items-start">
            <Link href={`/posts/${post.id}`} className="block underline">
              {post.title}
            </Link>
            {/* @ts-expect-error Server Component */}
            <DeleteForm id={post.id} />
          </div>
        ))}
      </section>
    </div>
  );
};
```

```
const DeleteForm = async ({ id }: { id: string }) => {
  const deletePost = async () => {
    "use server";
    await prisma.post.delete({ where: { id } });
    revalidatePath("/");
  };
  return (
    <form action={deletePost}>
      <button className="text-sm" type="submit">
        delete
      </button>
    </form>
  );
};
```

**Comment by icyJoseph** [comment]
I feel like I just saw, and tried to answer this... deja vú, but here it goes again:

What Next.js version have you got? It doesn't work for me, in either dev nor production mode.

I imagine the issue has to do with closure capturing, and how that would have been a way to leak secrets (it was actually shown to do that). [YouTube video](https://www.youtube.com/watch?v=fdKcbyEK66M)

For now I guess you gotta refactor to using a client component in the deletion bit:

```tsx
"use client";

import { deletePost } from "./actions";

export const DeleteForm = ({ id }: { id: string }) => {
  return (
    <form>
      <button
        className="text-sm"
        type="submit"
        formAction={() => deletePost({ id })}
      >
        delete
      </button>
    </form>
  );
};
```

Or! have a hidden unchanging field in that form, which contains the id to delete...

```tsx
import { deletePost } from "./actions";

export const DeleteForm = ({ id }: { id: string }) => {
  async function deleteAction(data: FormData) {
    "use server";

    const postId = data.get("postId");
    if (!postId) return;

    await deletePost({ id: postId.toString() });
  }

  return (
    <form action={deleteAction}>
      <button className="text-sm" type="submit">
        delete
      </button>
      {/* users can still change this value through the DevTools */}
      {/* it might be a good idea to hash it and un-hash it server side*/}
      <input name="postId" type="hidden" defaultValue={id} readOnly />
    </form>
  );
};
```

> I noticed a funny bug while doing this... If the hidden field is the first child of the form, then you get two input fields! https://github.com/vercel/next.js/issues/50087

Anyway, this has the downside that people could mess with the id on devTools and what not. I think in general you kind of don't want to leak the id to the client. You could hash the id, so that it makes it harder to "guess" other ids and potentially mess things up. Maybe I am just being too careful though...

Here's an example:

- encrypt/decrypt functions https://github.com/icyJoseph/next-battlesnake/blob/main/src/utils.ts
- here I encrypt every game uuid, and send that to the client instead https://github.com/icyJoseph/next-battlesnake/blob/main/src/pages/api/stats/index.ts#L36
- here, when a user wants to see more of a game, I take the encrypted uuid, decrypt it, and see if that exists on the db https://github.com/icyJoseph/next-battlesnake/blob/main/src/pages/api/stats/%5Buuid%5D.ts

If someone just tries to guess game id's they'll have a hard time actually generating one that, when decrypted is something that exists on my db.

**Reply by thanosoncode** [reply]
Haha yeah closed it to figure it out but then I thought boss Icy might want to answer this so re-opened 😄 

Hmm now it did not work in dev mode for me either. Interesting. 13.4.3. Anyway the point of this exercise was to do a crud using server components only that's why the whole DeleteForm shenanigans where I pass the post to it. Isn't it possible btw to do a crud server side?

Aren't all the ids already available to anyone? They're all there in each post's route

**Reply by icyJoseph** [reply]
Yeah, it's just me being over protective there, because if you know how the battle snakes work, I think you could try to invoke my snake with bad data which would pollute the DB 😅

Well using the hidden input you can do the delete "without js" but the closure over the if seems to no longer be possible, or was it ever? I tried to stay away from this since it's still alpha :/

**Comment by thanosoncode** [comment]
Could you explain why the id is not defined in the DeleteForm in my code?

**Reply by icyJoseph** [reply]
The way I understand it, it's because, for the client side that function is turned into an ID, and it's expected that its arguments and return value are serializable. That's why you label the function as `"use server"`

In the video I shared above, the closure was being captured and it allowed the leakage of secrets. My guess is that in the last versions, they nerfed the server action to prevent that. If you downgrade to 13.4.0 for example, does it work?

**Comment by thanosoncode** [comment]
Well that's interesting 😮 


https://github.com/vercel/next.js/assets/74588018/6ed1650d-a93a-44e6-a8e4-7ece80f70bc6

**Reply by icyJoseph** [reply]
Indeed! 

Alright so the closure capturing doesn't get the shorthand.

I tried, and failed, to do something like that, rename the input to postId, and then assigning like  `const id = postId`, but I did it at the component scope.

Could you try that too? 

Also, does it work on a production build?

I get a bit worried about learning an implementation trick at alpha stage, because the entire thing might just change tomorrow

**Comment by thanosoncode** [comment]
The explicit notation method like in the video works in production build (in 13.4.0) just fine!

As for the reassigning I'm not sure if I understood correctly but if you meant this
```
const DeleteForm = async ({ postId }: { postId: string }) => {
  const deletePost = async () => {
    "use server";
    const id = postId;
    await prisma.post.delete({ where: { id } });
    revalidatePath("/");
  };
  ```
  prisma thinks the id is optional for some reason 
  ```Error: Invalid `prisma.post.delete()` invocation:
{
  where: {
?   id?: String
  }
}```

**Reply by icyJoseph** [reply]
Mmm I'll test it out later on. Good that you found a fix! Still strange that the shorthand breaks it,👻

**Reply by thanosoncode** [reply]
I might start a new discussion cause this non shorthand in the action is kinda important. 
Thank you very much for your time

**Reply by icyJoseph** [reply]
I think maybe open an issue? You have a reproduction demo now.

**Reply by thanosoncode** [reply]
Done! https://github.com/vercel/next.js/issues/50445
